### PR TITLE
Add class to statically get asset URLs similar to Router::url() for routes.

### PR DESCRIPTION
**Pull request by ADmad**
Refs https://www.reddit.com/r/cakephp/comments/alnoz8/how_to_generate_an_asseturl_with_the_urlhelper_in/

### CODE REVIEW

**Comment by ADmad** [comment]
Once this is approved I can retrofit `UrlHelper` to use this class in it's methods and avoid code duplication.

One thing to note is this class only does URL encoding and doesn't do any HTML entities encoding. I believe the latter should be done only when this class is used through helpers where are meant to generate HTML.

File: src/View/Asset.php [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+use Cake\Core\Configure;
+use Cake\Core\Exception\Exception;
+use Cake\Core\Plugin;
+use Cake\Routing\Router;
+use Cake\Utility\Inflector;
+
+/**
+ * Class for generating asset URLs.
+ */
+class Asset
```
Comment (markstory) [inline src/View/Asset.php]: Should this be in Routing package?

File: src/View/Asset.php [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+    /**
+     * Inflect the theme name to its underscored version.
+     *
+     * @param string $name Name of the theme which should be inflected.
+     * @return string Inflected name of the theme
+     */
+    protected static function inflectThemeName(string $name): string
+    {
+        return Inflector::underscore($name);
```
Comment (markstory) [inline src/View/Asset.php]: Do we need this additional function call/indirection?

**Review by markstory — COMMENTED** [review]
If we add this class would we replace the guts of UrlHelper with this?

File: src/View/Asset.php (hunk 1 above)
Comment (ADmad) [reply src/View/Asset.php]: Probably, I wasn't too sure of the location.

File: src/View/Asset.php (hunk 2 above)
Comment (ADmad) [reply src/View/Asset.php]: I copied this from UrlHelper, I can remove it.

**Comment by ADmad** [comment]
> If we add this class would we replace the guts of UrlHelper with this?

Yes

File: src/View/Asset.php (hunk 1 above)
Comment (markstory) [reply src/View/Asset.php]: Routing seems like a better location for this if we want to make a general asset URL building class.

**Comment by ADmad** [comment]
- Moved `Asset` class under `Routing`
- Updated `UrlHelper` to use `Asset` class
- Added the ability to control the inflection type when inflecting theme/plugin name. I have kept default inflection to `underscore` for now but would like to change it to `dasherize` to consistent with Router.

File: src/Routing/Asset.php [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    protected static $inflectionType = 'underscore';
+
+    /**
+     * Set inflection type.
+     *
+     * @param string $inflectionType Inflection type
+     * @return void
+     * @psalm-param 'dasherize'|'underscore' $inflectionType
+     */
+    public static function setInflectionType(string $inflectionType)
```
Comment (dereuromark) [inline src/Routing/Asset.php]: you can set :void here as return type, right?

File: src/Routing/Asset.php [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+        return $request->getAttribute('webroot');
+    }
+
+    /**
+     * Splits a dot syntax plugin name into its plugin and filename.
+     * If $name does not have a dot, then index 0 will be null.
+     * It checks if the plugin is loaded, else filename will stay unchanged for filenames containing dot.
+     *
+     * @param string $name The name you want to plugin split.
+     * @return array Array with 2 indexes. 0 => plugin name, 1 => filename.
```
Comment (dereuromark) [inline src/Routing/Asset.php]: Isnt this string[] then?

File: src/Routing/Asset.php (hunk 3 above)
Comment (ADmad) [reply src/Routing/Asset.php]: yup, it's an oversight

File: src/Routing/Asset.php (hunk 4 above)
Comment (ADmad) [reply src/Routing/Asset.php]: No, since plugin can be null

**Comment by dereuromark** [comment]
Nice SOLID clean up :)

File: src/Routing/Asset.php [hunk 5]
```diff
@@ -26,17 +26,17 @@
...
      * @psalm-var 'dasherize'|'underscore'
      */
     protected static $inflectionType = 'underscore';
 
     /**
-     * Set inflection type.
+     * Set inflection type to use when inflecting plugin/theme name.
      *
-     * @param string $inflectionType Inflection type
+     * @param string $inflectionType Inflection type. Valid values: 'dasherize', 'underscore'.
```
Comment (dereuromark) [inline src/Routing/Asset.php]: I recommend making those class const then for clarity and to be easier used by dev.

Comment (ADmad) [reply src/Routing/Asset.php]: Theoretically one can use any `Inflector` method. Someone can use `'humanize'` if they prefer. I don't want to add constants for all `Inflector` methods. `RouterBuilder` methods also have `inflect` option for similar purpose.

I'll change the description to mention the value should be a valid `Inflector` method name.

Comment (dereuromark) [reply src/Routing/Asset.php]: Cool, ok.

**Comment by markstory** [comment]
Looks good :+1: Do you think this is a class we should document in the book? I can write something up in the routing chapter.

**Comment by ADmad** [comment]
@markstory Yes please, thanks.

**Pull request merged**